defrag/defrag.go

```diff
@@ -0,0 +1,276 @@
+	//
+	// It modifies the IPv4 layer in place, returning true if the
+	// modified layer is now a full IPv4 payload.
+	//
+	// If the passed-in IP layer is NOT fragmented, it will
+	// immediately return true without modifying the layer.
+	//
+	// If the IPv4 layer is a fragment and we don't have all
+	// fragments, it will return false and store whatever internal
+	// information it needs to eventually defrag the packet.
+	//
+	// If the IPv4 layer is the last fragment needed to reconstruct
+	// the packet, *ip will be set to the entire defragmented packet,
+	// and the function will return true.
+	DefragIPv4(IPv4 *layers.IPv4) (bool, error)
+
+	// DiscardOlderThan discards all packets we haven't seen a
+	// fragment for since 't'.
+	// It returns the number of packets discarded in this way.
+	DiscardOlderThan(t time.Time) int
+}
+
+const (
+	IPv4MinimumFragmentSize    = 576
+	IPv4MaximumSize            = 65535
+	IPv4MaximumFragmentOffset  = 8189
+	IPv4MaximumFragmentListLen = 8
+)
+
+// DefragIPv4 implements the DefragIPv4 interface.
+// It use a map of all the running flows
+func (d *IPv4Defragmenter) DefragIPv4(ip *layers.IPv4) (bool, error) {
+	// check if we need to defrag
+	if st := d.DontDefrag(ip); st == true {
+		return true, nil
+	}
+	// perfom security checks
+	st, err := d.SecurityChecks(ip)
+	if err != nil || st == false {
+		return st, err
+	}
+
+	// ok, got a fragment
+	debug.Printf("defrag: got ip.Id=%d ip.FragOffset=%d ip.Flags=%d\n",
+		ip.Id, ip.FragOffset*8, ip.Flags)
+
+	// do we already has seen a flow between src/dst with that Id
+	ipf := *NewIPv4Flow(ip)
+	var fl *FragmentList
+	var exist bool
+	d.Lock()
+	fl, exist = d.IPFlows[ipf]
+	if !exist {
+		debug.Printf("defrag: creating a new flow\n")
+		fl = new(FragmentList)
+		d.IPFlows[ipf] = fl
+	}
+	d.Unlock()
+	// insert, and if final build it
+	st, err = fl.Insert(ip)
+
+	// at last, if we hit the maximum frag list len
+	// without any defrag success, we just drop everything and
+	// raise an error
+	if st == false && fl.List.Len()+1 > IPv4MaximumFragmentListLen {
+		d.Lock()
+		fl = new(FragmentList)
+		d.IPFlows[ipf] = fl
+		d.Unlock()
+		return false, fmt.Errorf("defrag: Fragment List hits its maximum"+
+			"size(%d), without sucess. Flushing the list",
+			IPv4MaximumFragmentListLen)
+	}
+
+	return st, err
+}
+
+// DiscardOlderThan forgets FragmentList without any activity since
+// time t. It returns the number of FragmentList aka number of
+// fragment packets it has discarded.
+func (d *IPv4Defragmenter) DiscardOlderThan(t time.Time) int {
+	var nb int
+	d.Lock()
+	for k, v := range d.IPFlows {
+		if v.LastSeen.Before(t) {
+			nb = nb + 1
+			delete(d.IPFlows, k)
+		}
+	}
+	d.Unlock()
+	return nb
+}
+
+// DontDefrag returns true if the IPv4 packet do not need
+// any defragmentation
+func (d *IPv4Defragmenter) DontDefrag(ip *layers.IPv4) bool {
+	// don't defrag packet with DF flag
+	if ip.Flags&layers.IPv4DontFragment != 0 {
```

Comment:
At that point, I don't know if it's a fragment. It's in fact the first test I am doing when I received a IP layer... DF flag means "do nothing" for a defragmenter, it's not an error..